src/main/java/io/confluent/connect/jdbc/source/JdbcSourceTask.java

```diff
@@ -149,21 +151,35 @@ public void start(Map<String, String> properties) {
                 timestampColumns
             );
           }
-          partition = Collections.singletonMap(
-              JdbcSourceConnectorConstants.TABLE_NAME_KEY,
-              tableOrQuery
-          );
+          partitionsForTable(partitions, tableOrQuery);
           break;
         case QUERY:
           partition = Collections.singletonMap(
               JdbcSourceConnectorConstants.QUERY_NAME_KEY,
               JdbcSourceConnectorConstants.QUERY_NAME_VALUE
           );
+          partitions.add(partition);
           break;
         default:
           throw new ConnectException("Unexpected query mode: " + queryMode);
       }
-      Map<String, Object> offset = offsets == null ? null : offsets.get(partition);
+
+      Map<String, Object> offset = null;
+      int offsetProtocolVersion = 0;
+      if (offsets != null) {
+        for (Map<String, String> toCheckPartition : partitions) {
+          offset = offsets.get(toCheckPartition);
+          if (offset != null) {
+            log.trace("Found non-null offset {} for partition {}", offsets, toCheckPartition);
+            offsetProtocolVersion = Integer.parseInt(
+                toCheckPartition.getOrDefault(
+                    JdbcSourceConnectorConstants.OFFSET_PROTOCOL_VERSION,
+                    JdbcSourceConnectorConstants.PROTOCOL_VERSION_ZERO
+                ));
+            break;
+          }
+        }
+      }
```

Comment:
This logic is looking for a single protocol amongst all of the topics, yet it is breaking as soon as it finds the first. This might be easier to understand if we put that into a separate method, with JavaDoc that explains what it is doing and the assumptions it is making.

I'm not sure that the assumption about protocol versions might be valid. Imagine a connector runs with table A and B, and offsets are both stored. Then the connector is changed to use table A only and restarted. The offsets for table A and B both use protocol V0. Then the connector is upgraded to use protocol V1, and this updates the offsets for table A. Then the connector is changed to also consume from table B, but now the offsets for table A are V1 while the offsets for table B are still V1.

Is it not sufficient to always be able to read the offsets for each partition using that partition's protocol version? Obviously all future offsets would be written with the latest protocol version.

Reply:
Oh, maybe this is simply trying to find the latest protocol version to use for writes? We definitely need to extract this to a separate method and have good JavaDoc. 

Why not always write using the most recent protocol version?